enable bubble flow control for ring/torus deadlock avoidance

SeanNijjar opened the pull request:
Massive speedup on ring T3K (up to 32%) and considerable gain on 6u ring as well/

## Bubble Flow Control

This change adds bubble flow control, replacing the prior dateline VC approach for deadlock avoidance in rings/toruses in TT-Fabric. Bubble flow control is preferred here because it is far more buffering and compute efficient with our hardware.

Bubble flow control works for k-ary n-cube topologies - all of our supported "mesh" level topologies fall in this classification. It will not work for "intermesh" deadlock, hence is only a replacement for our dateline strategy. This flow control approach works by marking sender channels as "injection" and "forwarding" channels. 

Injection channels have special rules to follow when sending packets to their remote receiver channels (over Ethernet). Specifically, injection channels must ensure that the immediate downstream buffer has _atleast_ 2 buffer slots available before they can forward. Forwarding channels do not have this constraint.

#### Why 2 slots?

With two slots available, it is guaranteed that the ring can still make progress. If the injection channel forwards when there are two slots available, then atleast one other slot will still be available for the forwarding channel to send to. This means that even if every other receiver channel is full, atleast one packet can make forward progress. After this packet is forwarded, the upstream router will now have a free slot (a "bubble"), but note that no packet injection will be able to happen, only forwarding. The free slot propagates through the network ("bubbles up" - hence the name), until messages reach their final destination and start to drain out of the network.

Here are some diagrams to visualize the behaviour:

 ``` 
  ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────┐ 
  │  ┌─────────────────────────────────────────────────────────────────────────────────────────────────────┐  │ 
  │  │                                                                                                     │  │ 
  │  │                                                                                                     │  │ 
  │  │  ┌─────────────────────────────┐   ┌────────────────────────────┐  ┌─────────────────────────────┐  │  │ 
  │  │  │                             │   │                            │  │                             │  │  │ 
  │  │  │  ┌──┌──┌──┐                 │   │  ┌──┌──┌──┐                │  │  ┌──┌──┌──┐                 │  │  │ 
  │  └──┼─►│  │\/│\/├─────────────────┼───┼─►│\/│\/│\/├────────────────┼──┼─►│  │  │\/┼─────────────────┼──┘  │ 
  └─────┼─►│  │/\│/\│            ┌────┼───┼─►│/\│/\│/\│           ┌────┼──┼─►│  │  │/\│             ┌───┼─────┘ 
        │  └──└──└──┘            │    │   │  └──└──└──┘           │    │  │  └──└──└──┘             │   │       
        │                        │    │   │                       │    │  │                         │   │       
        │                        │    │   │                       │    │  │                         │   │       
        │         injector ──────┘    │   │        injector ──────┘    │  │            injector ────┘   │       
        │                             │   │                            │  │                             │       
        │                             │   │                            │  │                             │       
        └─────────────────────────────┘   └────────────────────────────┘  └─────────────────────────────┘       
 ```                                                                                                               
 Injectors can only inject when atleast 2 spaces are available. 
                                                                                                                
  
``` 
 ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────┐  
 │  ┌─────────────────────────────────────────────────────────────────────────────────────────────────────┐  │  
 │  │                                                                                                     │  │  
 │  │                                                                                                     │  │  
 │  │  ┌─────────────────────────────┐   ┌────────────────────────────┐  ┌─────────────────────────────┐  │  │  
 │  │  │                             │   │                            │  │                             │  │  │  
 │  │  │  ┌──┌──┌──┐                 │   │  ┌──┌──┌──┐                │  │  ┌──┌──┌──┐                 │  │  │  
 │  └──┼─►│\/│\/│\/├─────────────────┼───┼─►│\/│\/│\/├────────────────┼──┼─►│  │\/│\/┼─────────────────┼──┘  │  
 └─────┼─►│/\│/\│/\│            ┌────┼───┼─►│/\│/\│/\|           ┌────┼──┼─►│  │/\│/\│             ┌───┼─────┘  
       │  └──└──└──┘            │    │   │  └──└──└──┘           │    │  │  └──└──└──┘             │   │        
       │                        │    │   │                       │    │  │                         │   │        
       │                        │    │   │                       │    │  │                         │   │        
       │         injector ──────┘    │   │        injector ──────┘    │  │            injector ────┘   │        
       │                             │   │                            │  │                             │        
       │                             │   │                            │  │                             │        
       └─────────────────────────────┘   └────────────────────────────┘  └─────────────────────────────┘        
```
If only one slot is available, then only forwarding channels can push packets. In this case, only the second device can forward to the third to the free slot. Not that all other slots are full except one. Injectors are not allowed to inject/add traffic into the network. Therefore it is impossible to exhaust buffering in the network globally.

                                                                                                           
```
 ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────┐  
 │  ┌─────────────────────────────────────────────────────────────────────────────────────────────────────┐  │  
 │  │                                                                                                     │  │  
 │  │                                                                                                     │  │  
 │  │  ┌─────────────────────────────┐   ┌────────────────────────────┐  ┌─────────────────────────────┐  │  │  
 │  │  │                             │   │                            │  │                             │  │  │  
 │  │  │  ┌──┌──┌──┐                 │   │  ┌──┌──┌──┐  forward data  │  │  ┌──┌──┌──┐                 │  │  │  
 │  └──┼─►│\/│\/│\/├─────────────────┼───┼─►│  │\/│\/├────────────────┼──┼─►│\/│\/│\/┼─────────────────┼──┘  │  
 └─────┼─►│/\│/\│/\│            ┌────┼───┼─►│  │/\│/\|           ┌────┼──┼─►│/\│/\│/\│             ┌───┼─────┘  
       │  └──└──└──┘            │    │   │  └──└──└──┘           │    │  │  └─┬└──└──┘             │   │        
       │                        │    │   │   ▲                   │    │  │    │                    │   │        
       │                        │    │   │   │                   │    │  │    │                    │   │        
       │         injector ──────┘    │   │   │    injector ──────┘    │  │    │       injector ────┘   │        
       │                             │   │   │                        │  │    │                        │        
       │                             │   │   │                        │  │    │                        │        
       └─────────────────────────────┘   └───┼────────────────────────┘  └────┼────────────────────────┘        
                                             │                                │                                 
                                             │                                │                                 
                                             │          bubble moved          │                                 
                                             │             "up"               │                                 
                                             └────────────────────────────────┘                                 
 ```
After the second device forwards a packet, a slot opens on the second device, which the first device can forward into.
                                                                                                                
                                                                                                                
```
 ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────┐  
 │  ┌─────────────────────────────────────────────────────────────────────────────────────────────────────┐  │  
 │  │                                                                                                     │  │  
 │  │                                                                                                     │  │  
 │  │  ┌─────────────────────────────┐   ┌────────────────────────────┐  ┌─────────────────────────────┐  │  │  
 │  │  │                             │   │                            │  │                             │  │  │  
 │  │  │  ┌──┌──┌──┐  forward data   │   │  ┌──┌──┌──┐                │  │  ┌──┌──┌──┐                 │  │  │  
 │  └──┼─►│  │\/│\/├─────────────────┼───┼─►│\/|\/│\/├────────────────┼──┼─►│\/│\/│\/┼─────────────────┼──┘  │  
 └─────┼─►│  │/\│/\│            ┌────┼───┼─►│/\│/\│/\|           ┌────┼──┼─►│/\│/\│/\│             ┌───┼─────┘  
       │  └──└──└──┘            │    │   │  └─┬└──└──┘           │    │  │  └──└──└──┘             │   │        
       │    ▲                   │    │   │    │                  │    │  │                         │   │        
       │    │                   │    │   │    │                  │    │  │                         │   │        
       │    │    injector ──────┘    │   │    │   injector ──────┘    │  │            injector ────┘   │        
       │    │                        │   │    │                       │  │                             │        
       │    │                        │   │    │                       │  │                             │        
       └────┼────────────────────────┘   └────┼───────────────────────┘  └─────────────────────────────┘        
            │                                 │                                                                 
            │                                 │                                                                 
            │           bubble moved          │                                                                 
            │              "up"               │                                                                 
            └─────────────────────────────────┘                                                                 
```
And so on, the bubble keeps propagating through the network until messages drain out. At some point, there will again be 2 slots free in a device and new traffic can be injected.

## Notes on Some Structural Code Changes

There were a handful of slightly unrelated changes to the builder code to simplify some of the changes here. Specifically, some refactoring was done to the `FabricRouterBuilder` to restructure how channel mapping and behaviour tagging is done. 

In terms of general refactoring, the builder and channel mapping has started to more explicit have concepts of external facing and internal channels. External facing channels are those channels that are exposed to the outside of the logical router itself. Internal channels are those which are not exposed outside of the logical router, but instead are used for communication between cores that are cooperatively implementing the fabric router as a whole. For example, in Tensix mux mode, the channels on the erisc are internal only (as they are exclusively fed by the mux. To aid in this, some mapping functions have been created.

In terms of tagging,  the router builder will also decide on which channels are injection channels (or not) and pass that information down to the appropriate kernel builders (Tensix/Erisc).

In addition to the above, the changes also push the code towards (but not completely to) the below class hierarchy. The tensix mux builder configs are not quite matching the below description.

### Class Hierarchy
The builder class hierarchy is approaching the following description:

- `FabricRouterBuilder`: top level logical builder component, which instantiates and holds the erisc builder and zero or more of the tensix builders.
- Various builder configs: these are immutable initial configurations given to the router builder to understand how to build the underlying variant types
  - shared across multiple worker threads (not to be modified by the `FabricRouterBuilder` or its members
  - {FabricEriscDatamoverConfig, FabricTensixDatamoverConfig, FabricTensixDatamoverBaseConfig, FabricMuxConfig}
- Variant builders: responsible for generating all necessary state for constructing and launching the fabric kernel for a given core type (e.g. erisc router kernel), including setting up of compile and runtime args
  - Currently for some tensix builders, the `*Config` classes are actually setting up runtime/compile time args (but shouldn't be) 
  - {EriscDatamoverBuilder, FabricTensixDatamoverBuilder} 

## Speedup Summary (T3K)
These improvements are in addition to the up to 30% speedup from https://github.com/tenstorrent/tt-metal/pull/33192

### MESH
Packet Size: 4096, Geomean Speedup: 1.004110
Packet Size: 2048, Geomean Speedup: 1.006366
NType: NocSendType::NOC_UNICAST_SCATTER_WRITE, Geomean Speedup: 1.004662
NType: NocSendType::NOC_FUSED_UNICAST_ATOMIC_INC, Geomean Speedup: 1.005857
NType: NocSendType::NOC_UNICAST_WRITE, Geomean Speedup: 1.005193
Overall Geomean Speedup: 1.005237

### RING
Packet Size: 4096, Geomean Speedup: 1.313157
Packet Size: 2048, Geomean Speedup: 1.236920
NType: NocSendType::NOC_UNICAST_SCATTER_WRITE, Geomean Speedup: 1.311537
NType: NocSendType::NOC_FUSED_UNICAST_ATOMIC_INC, Geomean Speedup: 1.184692
NType: NocSendType::NOC_UNICAST_WRITE, Geomean Speedup: 1.332297
Overall Geomean Speedup: 1.274468

### LINEAR
Packet Size: 4096, Geomean Speedup: 1.002892
Packet Size: 2048, Geomean Speedup: 1.010865
NType: NocSendType::NOC_UNICAST_SCATTER_WRITE, Geomean Speedup: 1.006171
NType: NocSendType::NOC_FUSED_UNICAST_ATOMIC_INC, Geomean Speedup: 1.004311
NType: NocSendType::NOC_UNICAST_WRITE, Geomean Speedup: 1.008426
Overall Geomean Speedup: 1.006301

### Future Work
Cleanup remainder of VC1 specific code in fabric: [#33360](https://github.com/tenstorrent/tt-metal/issues/33360)

### Ticket
Closes [#32928](https://github.com/tenstorrent/tt-metal/issues/32928)

### Checklist
- [x] T3K Unit Tests: https://github.com/tenstorrent/tt-metal/actions/runs/19804749307 (same [as main](https://github.com/tenstorrent/tt-metal/actions/runs/19804170313))
- [x] T3K Fast: https://github.com/tenstorrent/tt-metal/actions/runs/19804752858
- [x] T3K Nightly: https://github.com/tenstorrent/tt-metal/actions/runs/19807024611
- [x] BH Multi-card Unit Tests: https://github.com/tenstorrent/tt-metal/actions/runs/19804754895 (all green except deskbox, with no runners. Previous runs green)
- [x] BH Multi-card Demo Tests: https://github.com/tenstorrent/tt-metal/actions/runs/19804755596
- [x] Galaxy Quick: https://github.com/tenstorrent/tt-metal/actions/runs/19806621429
- [x] Galaxy Unit: https://github.com/tenstorrent/tt-metal/actions/runs/19807076405 (same failure [on main](https://github.com/tenstorrent/tt-metal/actions/runs/19781211676))

SeanNijjar commented:
/codeowners ping

yugaoTT commented on tt_metal/fabric/fabric_context.cpp:
```diff
@@ -133,33 +131,28 @@ FabricContext::FabricContext(tt::tt_fabric::FabricConfig fabric_config) {
     // default router config don't care about the axis, since there's no optimization to it.
     this->router_config_ = get_edm_config_options(
-        tt::tt_fabric::FabricEriscDatamoverType::Default, tt::tt_fabric::FabricEriscDatamoverAxis::Short);
+        tt::tt_fabric::FabricEriscDatamoverAxis::Short);
```
I think we should be able to remove tt::tt_fabric::FabricEriscDatamoverAxis::Short, since there will be no difference between long and short axis now?

SeanNijjar replied:
Agreed. Will do in later PR.

SeanNijjar commented on tt_metal/fabric/erisc_datamover_builder.cpp:
```diff
@@ -919,24 +909,25 @@ std::vector<uint32_t> FabricEriscDatamoverBuilder::get_compile_time_args(uint32_
         this->fuse_receiver_flush_and_completion_ptr,
         fabric_context.need_deadlock_avoidance_support(this->direction_),
-        this->dateline_connection,
+        false,  // dateline_connection,
```
yes. I wanted to keep the names around for easier deletion on kernel side

SeanNijjar commented on tt_metal/fabric/fabric_context.cpp:
```diff
@@ -133,33 +131,28 @@ FabricContext::FabricContext(tt::tt_fabric::FabricConfig fabric_config) {
-        tt::tt_fabric::FabricEriscDatamoverType::Dateline, tt::tt_fabric::FabricEriscDatamoverAxis::Long);
+    this->dateline_router_config_[short_axis] = get_edm_config_options(tt::tt_fabric::FabricEriscDatamoverAxis::Short);
+    this->dateline_router_config_[long_axis] = get_edm_config_options(tt::tt_fabric::FabricEriscDatamoverAxis::Long);
 
```
I will leave an extra note with the cleanup issue tagged.

yugaoTT reviewed (commented):
oh sorry one thing I forgot to ask, should we also have the check for space change also in MUX mode?
In mux mode, we have channel 0 and 1 in line/ring, EWNS in mesh/torus, after this change, I think channel 0 should be checking for 2 slots available in ring, and worker channel + turn channel should be checking for 2 slots in torus?

SeanNijjar commented on tt_metal/fabric/erisc_datamover_builder_helper.cpp:
```diff
@@ -199,8 +170,8 @@ EdmLineFabricOpInterface::EdmLineFabricOpInterface(
                     src_curr_edm_config,
-                    build_in_worker_connection_mode,
-                    src_device_edm_type));
+                    std::vector<bool>(),  // Empty injection flags for legacy helper - not used
```
Note this whole function can be deprecated now that latency tests are enabled in the fabric test infrastructure (the only use of this path was the legacy fabric latency test)... That's all unrelated to this PR so this is just some placeholder change.

SeanNijjar commented on tt_metal/fabric/impl/kernels/edm_fabric/fabric_router_udm_mux_extension.cpp:
```diff
@@ -100,9 +109,17 @@ void forward_data(
+    bool is_injection_channel) {
     bool has_unsent_payload = get_ptr_val(my_channel_free_slots_stream_id.get()) != NUM_BUFFERS;
-    if (has_unsent_payload) {
+    bool send_packet = has_unsent_payload;
```
@yugaoTT is it feasible move `send_next_data` to some common header for reuse by both mux extensions? I found myself having to apply the same changes to both places.

SeanNijjar committed:
rebase conflict fix

SeanNijjar committed:
enable bubble flow control with first level acks

line working
ring working
mesh working

ring speedup: +27% geomean
ring latency +100ns (degradation), see issue #33318

Topology::Mesh, Packet Size: 4096, Geomean Speedup: 1.001777
Topology::Mesh, Packet Size: 2048, Geomean Speedup: 1.002850
Topology::Mesh, NType: NocSendType::NOC_UNICAST_SCATTER_WRITE, Geomean Speedup: 1.001896
Topology::Mesh, NType: NocSendType::NOC_FUSED_UNICAST_ATOMIC_INC, Geomean Speedup: 1.003510
Topology::Mesh, NType: NocSendType::NOC_UNICAST_WRITE, Geomean Speedup: 1.001535
Topology::Mesh, Overall Geomean Speedup: 1.002314
Topology::Ring, Packet Size: 4096, Geomean Speedup: 1.310491
Topology::Ring, Packet Size: 2048, Geomean Speedup: 1.233889
Topology::Ring, NType: NocSendType::NOC_UNICAST_SCATTER_WRITE, Geomean Speedup: 1.310414
Topology::Ring, NType: NocSendType::NOC_FUSED_UNICAST_ATOMIC_INC, Geomean Speedup: 1.180344
Topology::Ring, NType: NocSendType::NOC_UNICAST_WRITE, Geomean Speedup: 1.329377
Topology::Ring, Overall Geomean Speedup: 1.271613
Topology::Linear, Packet Size: 4096, Geomean Speedup: 1.002529
Topology::Linear, Packet Size: 2048, Geomean Speedup: 1.009027
Topology::Linear, NType: NocSendType::NOC_UNICAST_SCATTER_WRITE, Geomean Speedup: 1.006605
Topology::Linear, NType: NocSendType::NOC_FUSED_UNICAST_ATOMIC_INC, Geomean Speedup: 1.001991
Topology::Linear, NType: NocSendType::NOC_UNICAST_WRITE, Geomean Speedup: 1.007339
Topology::Linear, Overall Geomean Speedup: 1.005309

SeanNijjar committed:
potentially fully working now:
- Added injection channel identifiers for all sender channels
- Moved mux builders to have base of builder variant base
- Use injection channel identifiers to decide if bubble flow control is needed for the channel
- identify injection channels in top level router builder
- move more logic to top level router builder
- fix a bunch of bugs
- delete more VC1 code

SeanNijjar committed:
some cleanup and extra documentation

SeanNijjar committed:
add missing explicit include

SeanNijjar committed:
add quick ring deadlock sanity tests

SeanNijjar committed:
update targets

SeanNijjar committed:
update wh 6u targets

SeanNijjar committed:
fixed minor channel count bug with UDM fabric configs
